app/src/main/java/protect/card_locker/LoyaltyCardViewActivity.java

```diff
@@ -520,6 +533,79 @@ private void showInfoDialog() {
+        builder.setTitle(R.string.updateBalanceTitle);
+        FrameLayout container = new FrameLayout(this);
+        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(
+                ViewGroup.LayoutParams.MATCH_PARENT,
+                ViewGroup.LayoutParams.WRAP_CONTENT
+        );
+        params.leftMargin = 60;
+        params.rightMargin = 60;
+
+        LinearLayout layout = new LinearLayout(this);
+        layout.setOrientation(LinearLayout.VERTICAL);
+
+        TextView currentTextview = new TextView(this);
+        currentTextview.setText(getString(R.string.currentBalanceSentence, Utils.formatBalance(this, loyaltyCard.balance, loyaltyCard.balanceType)));
+        layout.addView(currentTextview);
+
+        TextView updateTextView = new TextView(this);
+        updateTextView.setText(getString(R.string.newBalanceSentence, Utils.formatBalance(this, loyaltyCard.balance, loyaltyCard.balanceType)));
+        layout.addView(updateTextView);
+
+        final EditText input = new EditText(this);
+        Context dialogContext = this;
+        input.setInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_DECIMAL);
+        input.setHint(R.string.updateBalanceHint);
+        input.addTextChangedListener(new SimpleTextWatcher() {
+            @Override
+            public void onTextChanged(CharSequence s, int start, int before, int count) {
+                BigDecimal newBalance;
+                try {
+                    newBalance = calculateNewBalance(loyaltyCard.balance, loyaltyCard.balanceType, s.toString());
+                } catch (ParseException e) {
+                    updateTextView.setTag(null);
+                    updateTextView.setError(getString(R.string.parsingBalanceFailed));
+                    updateTextView.setText(getString(R.string.newBalanceSentence, Utils.formatBalance(dialogContext, loyaltyCard.balance, loyaltyCard.balanceType)));
+                    return;
+                }
+
+                // Save new balance into this element
+                updateTextView.setTag(newBalance);
+                updateTextView.setError(null);
+                updateTextView.setText(getString(R.string.newBalanceSentence, Utils.formatBalance(dialogContext, newBalance, loyaltyCard.balanceType)));
+            }
+        });
+        layout.addView(input);
+        layout.setLayoutParams(params);
+        container.addView(layout);
+
+        builder.setView(container);
+        builder.setPositiveButton(R.string.ok, (dialogInterface, i) -> {
+            CharSequence balanceError = updateTextView.getError();
+            if (balanceError != null) {
+                Toast.makeText(getApplicationContext(), balanceError.toString(), Toast.LENGTH_SHORT).show();
+                return;
+            }
+
+            BigDecimal newBalance = (BigDecimal) updateTextView.getTag();
+            DBHelper.updateLoyaltyCardBalance(database, loyaltyCardId, newBalance);
+            this.onResume();
```

Comment:
Catima crashes if you press OK before you enter anything:

```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: me.hackerchick.catima.debug, PID: 8946
    java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String java.math.BigDecimal.toString()' on a null object reference
        at protect.card_locker.DBHelper.updateLoyaltyCardBalance(DBHelper.java:499)
        at protect.card_locker.LoyaltyCardViewActivity.lambda$showBalanceUpdateDialog$10$protect-card_locker-LoyaltyCardViewActivity(LoyaltyCardViewActivity.java:594)
        at protect.card_locker.LoyaltyCardViewActivity$$ExternalSyntheticLambda5.onClick(Unknown Source:4)
        at androidx.appcompat.app.AlertController$ButtonHandler.handleMessage(AlertController.java:167)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7870)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
I/Process: Sending signal. PID: 8946 SIG: 9
```

Can probably be fixed by checking tag instead of text:

```suggestion
            // Grab calculated balance from input field
            BigDecimal newBalance = (BigDecimal) input.getTag();
            if (newBalance == null) {
                return;
            }
            
            // Actually update balance
            DBHelper.updateLoyaltyCardBalance(database, loyaltyCardId, newBalance);

             // Reload UI
            this.onResume();
```

(Also removed the Toast because after testing more I realized it didn't make much sense given the new balance is always shown, so it should be pretty visible if the balance change is 0 due to it being invalid)

Reply:
Good catch! Not sure how I missed that.